12-offer-encoding.md

```diff
@@ -0,0 +1,769 @@
+revealing adjacent nodes in proofs (assuming there is a non-revealed TLV
+which has enough entropy).
+
+The Merkle tree's leaves are, in TLV-ascending order for each tlv:
+1. The H(`LnLeaf`,tlv).
+2. The H(`LnAll`||all-tlvs,tlv) where "all-tlvs" consists of all non-signature TLV entries appended in ascending order.
+
+The Merkle tree inner nodes are H(`LnBranch`, lesser-SHA256||greater-SHA256);
+this ordering means that proofs are more compact since left/right is
+inherently determined.
+
+If there are not exactly a power of 2 leaves, then the tree depth will
+be uneven, with the deepest tree on the lowest-order leaves.
+
+e.g. consider the encoding of an `offer` `signature` with TLVs TLV1, TLV2 and TLV3:
+
+```
+L1=H("LnLeaf",TLV1)
+L1nonce=H("LnAll"||TLV1||TLV2||TLV3,TLV1) 
+L2=H("LnLeaf",TLV2)
+L2nonce=H("LnAll"||TLV1||TLV2||TLV3,TLV2) 
+L3=H("LnLeaf",TLV3)
+L3nonce=H("LnAll"||TLV1||TLV2||TLV3,TLV3) 
+
+Assume L1 < L1nonce, and L2 > L2nonce.
+
+   L1    L1nonce                      L2   L2nonce                L3   L3nonce
+     \   /                             \   /                       \   /
+      v v                               v v                         v v
+L1A=H("LnBranch",L1||L1nonce) L2A=H("LnBranch",L2nonce||L2)  L3A=H("LnBranch",L3nonce||L3)
+                 
+Assume L1A < L2A:
+
+       L1A   L2A                                 L3A=H("LnBranch",L3nonce||L3)
+         \   /                                    |
+          v v                                     v
+  L1A2A=H("LnBranch",L1A||L2A)                   L3A=H("LnBranch",L3nonce||L3)
+  
+Assume L1A2A > L3A:
+
+  L1A2A=H("LnBranch",L1A||L2A)          L3A
+                          \            /
+                           v          v
+                Root=H("LnBranch",L3A||L1A2A)
+
+Signature = SIG("lightningoffersignature", Root, nodekey)
+```
+
+# Offers
+
+Offers are a precursor to an invoice: readers will either request an invoice
+(or multiple) or send an invoice based on the offer.  An offer can be much longer-lived than a
+particular invoice, so has some different characteristics; in particular it
+can be recurring, and the amount can be in a non-lightning currency.  It's
+also designed for compactness, to easily fit inside a QR code.
+
+The human-readable prefix for offers is `lno`.
+
+## TLV Fields for Offers
+
+1. `tlv_stream`: `offer`
+2. types:
+    1. type: 2 (`chains`)
+    2. data:
+        * [`...*chain_hash`:`chains`]
+    1. type: 6 (`currency`)
+    2. data:
+        * [`...*utf8`:`iso4217`]
+    1. type: 8 (`amount`)
+    2. data:
+        * [`tu64`:`amount`]
+    1. type: 10 (`description`)
+    2. data:
+        * [`...*utf8`:`description`]
+    1. type: 12 (`features`)
+    2. data:
+        * [`...*byte`:`features`]
+    1. type: 14 (`absolute_expiry`)
+    2. data:
+        * [`tu64`:`seconds_from_epoch`]
+    1. type: 16 (`paths`)
+    2. data:
+        * [`...*blinded_path`:`paths`]
```

Comment:
pluralization mismatch: `blinded_path` vs `paths`. If this is supposed to represent an individual entry of the `paths` array, `path` should probably be singular in this line.

Reply:
There is no mismatch here, the syntax is not obvious but `...*blinded_path` means an array of blinded paths.